### ISSUE TITLE
Game does not handle invalid player position

### ISSUE THREAD

**Issue by Spikeone** [Return-To-The-Roots/s25client#935]
If player 1 has an invalid start position, the problem is shown in the console but not handled properly. The game keeps showing the loading screen.

![image](https://user-images.githubusercontent.com/1276342/48371278-c13be180-e6bb-11e8-8b08-b577782eb467.png)

Map: Campaign Miss200.wld

**Comment by Flamefire** [comment]
Ok I found the issue. What happens:
- Game start
- switch to loading screen desktop (enqueued for next draw call)
- game loading starts
- failure -> loading screen gets notified -> shows msgbox with the failure (closing this would go back to menu)
- desktop switch executed, all open windows get closed
- Result: Loading desktop shown expecting the msgbox to return but it doesn't

So the problem is a more general one: Opening a new window while a desktop switch is underway does not work unless the function (showAfterSwitch) is used which was introduced as a workaround for issues like this. However it requires the desktop to know, that it is switched (from/to).

Ideas:
1. If a desktop switch is under way when calling `Show(window)` the window is shown after the switch
  - Possible problem: This window might be intended for the current desktop
2. `Show` takes the desktop on which the window is shown as a param (or the window constructor requires the parent to be set to the parent desktop) and the window will get shown and closed with its desktop
  - Will be quite some work 
  - Maybe problem for windows opening windows (e.g. help windows from buildings) but they can open it for the same desktop (they know their parent)
  - Lua system would also need to know the current desktop. Not sure if this coupling is great...

2. seems to be the best soultion though. @Flow86 What do you think?

**Comment by Flamefire** [comment]
@Spikeone How does S2 treat the case where there are N players but not all have start positions?

**Comment by Spikeone** [comment]
@Flamefire The game trusts the editor - the editor sets a flag either 1 = valid, or 0 = invalid. https://settlers2.net/documentation/world-map-file-format-wldswd/

The editor always tested start positions when saving (BQ I guess). If anything failed, you got a window showing:
![image](https://user-images.githubusercontent.com/1276342/83942281-79773080-a7f2-11ea-9325-d7f88baec0d5.png)

A world was valid when you got "World is O.K."
![image](https://user-images.githubusercontent.com/1276342/83942295-8bf16a00-a7f2-11ea-8dab-51d5fcd4a661.png)

I'd not test what happens if you use an invalid world in Settlers 2 as they surely thought this was impossible with their editor (not regarding altered maps).

**Comment by Flamefire** [comment]
So what about that `Miss200.wld` you reported here? Is that marked invalid? Does it load in S2? Where is the player 1 (I guess 2nd player) HQ?

**Comment by Spikeone** [comment]
It's a *.wld, those can't be used in endless play. I don't know how much the game relies on the RTX files.

Renaming it to swd (so the original editor loads it) results in:
![image](https://user-images.githubusercontent.com/1276342/83949856-6b430780-a826-11ea-8c49-821f6807c2d6.png)

The game loads the map with 7 players, just as RttR does:
![image](https://user-images.githubusercontent.com/1276342/83949880-99c0e280-a826-11ea-872c-c24ca8b1fef2.png)

Game also starts as it should:
![image](https://user-images.githubusercontent.com/1276342/83949889-b6f5b100-a826-11ea-81d7-ab6b02757df0.png)

This also applies to the campaign itself, there are 7 players listed:
![image](https://user-images.githubusercontent.com/1276342/83949962-60d53d80-a827-11ea-86cc-32ec3483215e.png)

Overall I remember the following: If it was not possible to place the player HQ on startup, it simply did not spawn. I once (years ago) had a map where player positions were quite close to each other. The map was flagged valid but when starting the game a HQ was missing as it was surrounded by other, close players.

So regarding this issue, the game silently failed to place some players on startup.

**Comment by Flow86** [comment]
I think the point was, that you can't trust the count of players in the map headers - you always had to check a player hq position was set

**Comment by Flamefire** [comment]
> you can't trust the count of players in the map headers

But the whole point of the header is to not need to parse the whole file which takes much longer.

So the question is: What shall we do with invalid maps? The current handling was to reject such invalid maps. This is similar to other files: If the header does not match the file contents the file is invalid and is rejected on loading. The bug here is due to the design the message box is not show and game is not aborted once that invalid condition is detected.

I'm surprised the original game ignored this. So what shall we do? Abort as we currently (intend to) do or do the same strange thing the original does and simply don't place a HQ (but log a message)

**Comment by Flow86** [comment]
yes we should reject them. since we focus on MP here, it should be essential that a HQ can be placed.

Perhaps we can add some special check-code for the map files to verify the header? (its only the player count vs hq-positions we need to check)

**Comment by Flamefire** [comment]
Ok, repeating my proposal from https://github.com/Return-To-The-Roots/s25client/pull/1248#discussion_r436256554 here before it gets lost:

> BUT: Maybe we should validate that all players are set. For that I'd move the Point class to libutil and let an s2map in libsiedler2 return an array with player positions for each player in the map. This allows to validate in libsiedler2 that each player has a unique start position and in s25client that there are as many start positions as in the header and each is set. I'd do the latter in s25client because it might be that we want players without a start position.

**Comment by DevOpsOfChaos** [comment]
I checked current `master`.

The original loading-screen hang seems likely to be addressed already: `dskGameLoader::ShowErrorMsg()` now uses `WINDOWMANAGER.ShowAfterSwitch()` when the loader desktop is not active, which matches the desktop-switch race described above.

Also, `MapLoader::PlaceHQs()` currently rejects used players without a valid HQ position by returning `false`.

So the remaining part, if any, seems to be the deeper map/header validation question discussed above rather than the original loading-screen hang itself. Unless this can still be reproduced on current `master`, this may need either retesting or a narrower follow-up issue for validating header player counts against actual start positions.
